File: interpreter/c-sharp/JobManager.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Interpreter
{
    class JobManager
    {
        public List<IJobExpression> CreateExpressionTree(string formula)
        {
            List<IJobExpression> tree = new List<IJobExpression>();

            foreach (char role in formula)
            {
                switch (role)
                {
                    case 'G':
                        tree.Add(new DeveloperExpression());
                        break;
                    case 'M':
                        tree.Add(new AccountingExpression());
                        break;
                    case 'L':
                        tree.Add(new TeamLeadExpression());
                        break;
                    default:
                        throw new Exception("Unexpected role!");
                }
            }
            return tree;
        }

        public void RunExpression(Context context)
        {
            foreach (IJobExpression expression in CreateExpressionTree(context.formula))
            {
                expression.Interpret(context);
            }
        }
    }
}

File: interpreter/go/job-manager.go
package main

import "errors"

type JobManager struct {
}

func NewJobManager() *JobManager {
	return &JobManager{}
}

func (jm JobManager) createExpressionTree(formula string) ([]JobExpression, error) {
	tree := make([]JobExpression, 0)

	for _, role := range formula {
		switch role {
		case 'G':
			tree = append(tree, NewDeveloperExpression())
		case 'M':
			tree = append(tree, NewAccountingExpression())
		case 'L':
			tree = append(tree, NewTeamLeadExpression())
		default:
			return nil, errors.New("Unexpected role " + string(role))
		}
	}

	return tree, nil
}

func (jm JobManager) RunExpression(ctx *Context) error {
	tree, err := jm.createExpressionTree(ctx.Formula)
	if err != nil {
		return err
	}
	for _, expression := range tree {
		expression.Interpret(ctx)
	}
	return nil
}
